### ISSUE TITLE
Exist validation error after updating on Yii 2.0.8

### ISSUE THREAD

**Issue by Tigrov** [yiisoft/yii2#11467]
Steps to reproduce the problem:
It needs two models, for example ```Request``` and ```Address```:
```php
/**
 * ...
 * @property integer $user_id
 * @property integer $shipping_address_id
 * ...
 */
class Request extends ActiveRecord
{
    ...
    public function rules()
    {
        return [
            ...
            [['!shipping_address_id'], 'exist', 'targetClass' => Address::className(),
                'targetAttribute' => 'id', 'filter' => ['user_id' => $this->user_id]],
            ...
        ];
    }
    ...
}
```

```php
/**
 * ...
 * @property integer $id
 * @property integer $user_id
 * @property string $email
 * ...
 */
class Address extends ActiveRecord
{
    ...
}
```
And controller:
```php
...
public function actionCreate()
{
    ...
    $request = new Request;
    $address = new Address;
    // Load all models
    $isLoaded = $request->load(Yii::$app->request->post());
    $isLoaded = $address->load(Yii::$app->request->post()) && $isLoaded;
    if ($isLoaded) {
        // If User is guest, create new User using $address->email as username
        $user = Yii::$app->user->isGuest 
            ? User::create($address->email) 
            : Yii::$app->user->identity;
        $address->user_id = $user->id;
        if ($address->save()) {
            $request->user_id = $address->user_id;
            $request->shipping_address_id = $address->id;
            if ($request->save()) {
                // redirect
            }
        }
    }
    ...
    return $this->render('create', ['request' => $request, 'address' => $address]);
}
...
```
After post a form I get an error for ```Request``` model: "Shiping Address is invalid."

I expected that it will validate the ```Request``` model when calling ```$request->save()``` and will use ```Request::rules()``` in this moment.
But insted it the ```rules()``` cached by ```$request->load()``` when this process calls the stack ```Model::setAttributes()``` -> ```Model::safeAttributes()``` -> ```Model::scenarios()``` -> ```Model::getValidators()``` and caches the validators there.
For this reason, ```$this->user_id``` is equal ```null``` in ```Request::rules()``` (```'filter' => ['user_id' => null]```) when calling ```$request->save()```, but must be the same as ```$request->user_id``` at this moment.

Yii version: 2.0.8
PHP version: 5.6.14

**Comment by Faryshta** [comment]
Thats an expected behavior, the strange is that it worked at 2.0.7. When you invoke `load()` the model creates the validators BEFORE assigning the values to the attributes since it requires to know which attributes are safe to be loaded.

Then when you create the validator `$this->user_id` is `null`. My advice is to change it like this

```php
'filter' => function ($query) {
    $query->andWhere(['user_id' => $this->user_id]);
}
```

Also consider using transactionw when dealing with more than one model

**Comment by Tigrov** [comment]
@Faryshta Thanks, it's solving my issue.

I found the problem, this rule was not executing at 2.0.7 and lower. Becaus there was a problem with unsafe attributes ```['!shipping_address_id']``` that was solved in 2.0.8 #11153.

**Comment by Faryshta** [comment]
close this issue then :+1:
